crates/dojo-core/src/database_test.cairo

```diff
@@ -113,17 +112,86 @@ fn test_database_del() {
 #[test]
 #[available_gas(10000000)]
 fn test_database_scan() {
-    let even = array![2, 4].span();
-    let odd = array![1, 3].span();
-    let layout = array![251, 251].span();
+    let even = array![2, 4, 6].span();
+    let odd = array![1, 3, 5].span();
+    let layout = array![251, 251, 251].span();
 
-    set_with_index('table', 'even', 0, even, layout);
-    set_with_index('table', 'odd', 0, odd, layout);
+    set_with_index('table', 'even', array!['x'].span(), 0, even, layout);
+    set_with_index('table', 'odd', array!['x'].span(), 0, odd, layout);
 
-    let (keys, values) = scan('table', Option::None(()), 2, layout);
-    assert(keys.len() == 2, 'Wrong number of keys!');
+    let values = scan(Clause::All('table'), 3, layout);
     assert(values.len() == 2, 'Wrong number of values!');
-    assert(*keys.at(0) == 'even', 'Wrong key at index 0!');
+    (*(*values.at(0)).at(0)).print();
     assert(*(*values.at(0)).at(0) == 2, 'Wrong value at index 0!');
     assert(*(*values.at(0)).at(1) == 4, 'Wrong value at index 1!');
+    assert(*(*values.at(0)).at(2) == 6, 'Wrong value at index 1!');
+
+    let where = MemberClause { model: 'table', member: 'x', value: 2 };
+
+    let values = scan(Clause::Member(where), 32, layout);
+    assert(values.len() == 1, 'Wrong number of values clause!');
+}
+
+#[test]
+#[available_gas(10000000)]
+fn test_database_scan_where() {
+    let some = array![1, 4].span();
+    let same = array![1, 3].span();
+    let other = array![5, 5].span();
+    let layout = array![251, 251].span();
+
+    set_with_index('table', 'some', array!['p', 'x'].span(), 0, some, layout);
+    set_with_index('table', 'same', array!['p', 'x'].span(), 0, same, layout);
+    set_with_index('table', 'other', array!['p', 'x'].span(), 0, other, layout);
+
+    let values = scan(Clause::All('table'), 2, layout);
+    assert(values.len() == 3, 'Wrong number of values!');
+    assert(*(*values.at(0)).at(0) != 0, 'value is not set');
+
+    let mut where = MemberClause { model: 'table', member: 'x', value: 5 };
+
+    let values = scan(Clause::Member(where), 2, layout);
+    assert(values.len() == 1, 'Wrong len for x = 5');
+    assert(*(*values.at(0)).at(0) == 5, 'Wrong value 0 for x = 5');
+    assert(*(*values.at(0)).at(1) == 5, 'Wrong value 1 for x = 5');
+
+    where.value = 4;
+    let values = scan(Clause::Member(where), 2, layout);
+    assert(values.len() == 1, 'Wrong len for x = 1');
+
+    where.value = 6;
+    let values = scan(Clause::Member(where), 2, layout);
+    assert(values.len() == 0, 'Wrong len for x = 6');
+}
+
+#[test]
+#[available_gas(20000000)]
+fn test_database_scan_where_deletion() {
```

Comment:
@matzayonc @neotheprogramist this test is failing but should pass. we aren't removing the member index when deleting a entity